File: Tests/vs_cpp/vs_cpp/CS_Handler.cpp
//#include "precompile.h"
# include "CS.h"
# include "xxlib.h"
using namespace xxlib;

bool PackageHandle( ByteBuffer& bb )
{
    short pkgId = -1; 
    if( !bb.Read( pkgId ) )
    {
        return false;
    }

    // custom pkgId handler code here

    switch( pkgId )
    {
    case 0:
    {
        ::CS::RootPkg1 o;
        if( !bb.Read( o ) )
        {
            return false;
        }
            
        // code here

        return true;
    }
    case 1:
    {
        ::CS::Ns2::Pkg2 o;
        if( !bb.Read( o ) )
        {
            return false;
        }
            
        // code here

        return true;
    }
    case 2:
    {
        ::CS::Ns1::Pkg1 o;
        if( !bb.Read( o ) )
        {
            return false;
        }
            
        // code here

        return true;
    }
    case 3:
    {
        ::CS::Ns1::Pkg2 o;
        if( !bb.Read( o ) )
        {
            return false;
        }
            
        // code here

        return true;
    }
    default:
        return false;
    }
}

bool PackageHandle( char* buf, int len )
{
    ByteBuffer bb( buf, len, len, true );
    return PackageHandle( bb );
}

File: Tests/vs_cs/vs_cs/CS_Handler.cs

using System;
using System.Collections.Generic;
using xxlib;
using CS;

public static class PackageHandler
{
    static bool Handle( ByteBuffer bb )
    {
        short pkgId = 0;
        try
        {
            bb.Read( ref pkgId );
        }
        catch
        {
            return false;
        }

        // custom pkgId handler code here

        switch( pkgId )
        {
        case global::CS.RootPkg1.packageId:
            {
                var o = new global::CS.RootPkg1();
                try
                {
                    bb.Read( ref o );
                }
                catch
                {
                    return false;
                }

                // code here

                return true;
            }
        case global::CS.Ns2.Pkg2.packageId:
            {
                var o = new global::CS.Ns2.Pkg2();
                try
                {
                    bb.Read( ref o );
                }
                catch
                {
                    return false;
                }

                // code here

                return true;
            }
        case global::CS.Ns1.Pkg1.packageId:
            {
                var o = new global::CS.Ns1.Pkg1();
                try
                {
                    bb.Read( ref o );
                }
                catch
                {
                    return false;
                }

                // code here

                return true;
            }
        case global::CS.Ns1.Pkg2.packageId:
            {
                var o = new global::CS.Ns1.Pkg2();
                try
                {
                    bb.Read( ref o );
                }
                catch
                {
                    return false;
                }

                // code here

                return true;
            }
        default:
            return false;
        }
    }
}

